packages/@lwc/engine-core/src/framework/rendering.ts

```diff
@@ -255,18 +254,23 @@ function mountCustomElement(
     renderer: RendererAPI
 ) {
     const { sel, owner } = vnode;
-    const UpgradableConstructor = getUpgradableConstructor(sel, renderer);
+    const { getUpgradableElement, getUserConstructor } = renderer;
     /**
      * Note: if the upgradable constructor does not expect, or throw when we new it
      * with a callback as the first argument, we could implement a more advanced
      * mechanism that only passes that argument if the constructor is known to be
      * an upgradable custom element.
      */
     let vm: VM | undefined;
-    const elm = new UpgradableConstructor((elm: HTMLElement) => {
+
+    const upgradeCallback = (elm: HTMLElement) => {
         // the custom element from the registry is expecting an upgrade callback
         vm = createViewModelHook(elm, vnode, renderer);
-    });
+    };
+    const UpgradableConstructor = getUpgradableElement(sel);
+    const UserConstructor = getUserConstructor(upgradeCallback);
```

Comment:
this worries me a little bit... core should not know much about this since it doesn't know about pivots, or anything like that. Imagine the case where all elements MUST be registered as globals, and the renderer implements a simple getUpgradableElement() who's job is simply to get the Ctor from the registry and return it here.

Reply:
Yeah, I could imagine this being refactored. It's a bit messy, but my goal was to get it to work across both engine-server and engine-dom (with native CEs and without), which had slightly different needs.

Reply:
OK after discussion, we landed on something like:

```js
const elm = renderer.createCustomElement(tagName, upgradeCallback)
```